alica_engine/src/engine/Assignment.cpp

```diff
@@ -1,532 +1,448 @@
-    if (this->epRobotsMapping->getSize() != otherAssignment->epRobotsMapping->getSize()) {
-        return false;
-    }
-    // check for same entrypoints
-    for (int i = 0; i < this->epRobotsMapping->getSize(); ++i) {
-        if (this->epRobotsMapping->getEp(i)->getId() != otherAssignment->epRobotsMapping->getEp(i)->getId()) {
-            return false;
-        }
-    }
-    // check for same robots in entrypoints
-    for (short i = 0; i < this->epRobotsMapping->getSize(); ++i) {
-        if (this->epRobotsMapping->getRobots(i)->size() != otherAssignment->epRobotsMapping->getRobots(i)->size()) {
-            return false;
-        }
-
-        for (auto& robot : *(this->epRobotsMapping->getRobots(i))) {
-            auto iter = find_if(otherAssignment->epRobotsMapping->getRobots(i)->begin(), otherAssignment->epRobotsMapping->getRobots(i)->end(),
-                                [&robot](const supplementary::AgentID* id) { return *robot == *id; });
-            if (iter == otherAssignment->epRobotsMapping->getRobots(i)->end()) {
-                return false;
+    for (int i = 0; i < static_cast<int>(_assignmentData.size()); ++i) {
+        const bool isTargetEp = ep->getIndex() == i;
+        if (isTargetEp) {
+            const State* s = ep->getState();
+            for (AgentIDConstPtr id : agents) {
+                auto it = std::find_if(_assignmentData[i].begin(), _assignmentData[i].end(), [id](AgentStatePair asp) { return asp.first == id; });
+                if (it == _assignmentData[i].end()) {
+                    _assignmentData[i].emplace_back(id, s);
+                } else {
+                    it->second = s;
+                }
             }
+        } else {
+            _assignmentData[i].removeAllIn(agents);
         }
     }
-    return true;
 }
 
-/**
- * Test whether at least one robot is working on a task or succeeded with a task.
- * @param ep An EntryPoint identifying the task in question.
- * @return bool
- */
-bool Assignment::isEntryPointNonEmpty(const EntryPoint* ep) const
+void Assignment::adaptTaskChangesFrom(const Assignment& as)
 {
-    auto r = this->epRobotsMapping->getRobotsByEp(ep);
-    if (r != nullptr && r->size() > 0) {
-        return true;
+    if (as.getPlan() != getPlan()) {
+        *this = as;
+        return;
+    }
+    const int epCount = _assignmentData.size();
+    for (int i = 0; i < epCount; ++i) {
+        AgentStatePairs n = as._assignmentData[i];
+        for (AgentStatePair& asp : n) {
+            const State* s = _assignmentData[i].getStateOfAgent(asp.first);
+            if (s) {
+                asp.second = s;
+            }
+        }
+        _assignmentData[i] = std::move(n);
     }
-    auto epSuc = this->epSucMapping->getRobots(ep);
-    return (epSuc != nullptr && epSuc->size() > 0);
 }
 
-bool Assignment::updateRobot(const supplementary::AgentID* robot, const EntryPoint* ep, const State* s)
+bool Assignment::removeAllIn(const AgentGrp& limit, const State* watchState)
```

Comment:
The names in this interface are not very clear to me. What are limit and watchState?

Reply:
remove all Agents in limit and return of watchstate was touched. Do you have a better suggestion?